napari/_vispy/layers/surface.py

```diff
@@ -43,53 +48,85 @@ def __init__(self, layer) -> None:
         self._on_data_change()
 
     def _on_data_change(self):
-        ndisplay = self.layer._slice_input.ndisplay
-        if len(self.layer._data_view) == 0 or len(self.layer._view_faces) == 0:
-            vertices = None
-            faces = None
-            vertex_values = np.array([0])
-        else:
+        vertices = None
+        faces = None
+        vertex_values = None
+        vertex_colors = None
+        if len(self.layer._data_view) and len(self.layer._view_faces):
             # Offsetting so pixels now centered
             # coerce to float to solve vispy/vispy#2007
             # reverse order to get zyx instead of xyz
             vertices = np.asarray(
                 self.layer._data_view[:, ::-1], dtype=np.float32
             )
-            # due to above xyz>zyx, also reverse order of faces to fix handedness of normals
+            # due to above xyz>zyx, also reverse order of faces to fix
+            # handedness of normals
             faces = self.layer._view_faces[:, ::-1]
-            vertex_values = self.layer._view_vertex_values
-
-        if vertices is not None and ndisplay == 3 and self.layer.ndim == 2:
-            vertices = np.pad(vertices, ((0, 0), (0, 1)))
-
-        # manually detach filters when we go to 2D to avoid dimensionality issues
-        # see comments in napari#3475. The filter is set again after set_data!
-        if ndisplay == 2:
-            filt = self.node.shading_filter
-            try:
-                self.node.detach(filt)
-                self.node.shading = None
-                self.node.shading_filter = None
-            except ValueError:
-                # sometimes we try to detach non-attached filters, which causes a ValueError
-                pass
+
+            values = self.layer._view_vertex_values
+            if len(values):
+                vertex_values = values
+
+            colors = self.layer._view_vertex_colors
+            if len(colors):
+                vertex_colors = colors
+
+        # making sure the vertex data is 3D prevents shape errors with
+        # attached filters, instead of trying to attach/detach each time
+        if vertices is not None and vertices.shape[-1] == 2:
+            vertices = np.pad(
+                vertices,
+                ((0, 0), (0, 1)),
+                mode='constant',
+                constant_values=0,
+            )
+        assert vertices is None or vertices.shape[-1] == 3
```

Comment:
I have vague memories about this (always using 3d vertices) being a problem, great to see you solved it instead of the workaround we had :)